lib/core/unittest/CConcurrencyTest.cc

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#include "CConcurrencyTest.h"
+
+#include <core/CLogger.h>
+#include <core/Concurrency.h>
+
+#include <numeric>
+
+using namespace ml;
+
+void CConcurrencyTest::testAsyncWithExecutors() {
+
+    core::stopDefaultAsyncExecutor();
+
+    std::string tags[]{"sequential", "parallel"};
+
+    for (std::size_t t = 0; t < 2; ++t) {
+        LOG_DEBUG(<< "Testing " << tags[t]);
+
+        auto result =
+            core::async(core::defaultAsyncExecutor(), []() { return 42; });
+        CPPUNIT_ASSERT_EQUAL(42, result.get());
+
+        result = core::async(core::defaultAsyncExecutor(),
+                             [](int i) { return i; }, 43);
+        CPPUNIT_ASSERT_EQUAL(43, result.get());
+
+        result = core::async(core::defaultAsyncExecutor(),
+                             [](int i, int j) { return i + j; }, 22, 22);
+        CPPUNIT_ASSERT_EQUAL(44, result.get());
+
+        core::startDefaultAsyncExecutor();
+    }
+
+    core::stopDefaultAsyncExecutor();
+}
+
```

Comment:
would be nice to have a test case that shows an error situation, a task that simply throws. I also wonder what `result` should provide in this case, e.g. there could be `result.success()` returning `true` or `false` and `result.error()` which gives back the error message.

Reply:
Yes I was reflecting that I should also add a thread pool test, because even though it is being tested indirectly in quite a few places I also want to check the balancing fallbacks work properly  with very uneven task durations. I'll beef up the testing a bit.

Reply:
I added [testing](https://github.com/elastic/ml-cpp/pull/322/commits/780ebb0177036c0ba47990fc9cf44283ad7e8406) along the lines you suggest. Interestingly, it shows up an undesirable feature of futures which is there appears to be no way to tell if an exception has been caught with std::future<void> because you can't call get and this is the only method which propagates this information. I've tweaked the parallel_for_each functions accordingly. (Of course we shouldn't be throwing and catching exceptions locally, but this tidies up a loose end.)